porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go

```diff
@@ -162,82 +209,99 @@ func (r *PackageVariantReconciler) ensurePackageVariant(ctx context.Context,
 		},
 	}
 
+	klog.Infoln(fmt.Sprintf("package variant %q is creating package revision %q", pv.Name, newPR.Name))
 	if err := r.Client.Create(ctx, newPR); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (r *PackageVariantReconciler) findAndUpdateExistingRevision(ctx context.Context,
+func (r *PackageVariantReconciler) findAndUpdateExistingRevisions(ctx context.Context,
 	pv *api.PackageVariant,
 	upstream *porchapi.PackageRevision,
-	prList *porchapi.PackageRevisionList) (*porchapi.PackageRevision, error) {
-	// First, check if a downstream package exists. If not, just return nil. The
-	// caller will create one.
-	downstream := r.getDownstreamPR(pv, prList)
-	if downstream == nil {
+	prList *porchapi.PackageRevisionList) ([]*porchapi.PackageRevision, error) {
+	downstreams := r.getDownstreamPRs(ctx, pv, prList)
+	if downstreams == nil {
+		// If there are no existing target downstream packages, just return nil. The
+		// caller will create one.
 		return nil, nil
 	}
-
-	// Determine if the downstream package needs to be updated. If not, return
-	// the downstream package as-is.
-	if r.isUpToDate(pv, downstream) {
-		return downstream, nil
-	}
-
-	if porchapi.LifecycleIsPublished(downstream.Spec.Lifecycle) {
-		var err error
-		downstream, err = r.copyPublished(ctx, downstream, pv)
+	var err error
+	for i, downstream := range downstreams {
+		if r.isUpToDate(pv, downstream) {
+			continue
+		}
+		if porchapi.LifecycleIsPublished(downstream.Spec.Lifecycle) {
+			downstream, err = r.copyPublished(ctx, downstream, pv, prList)
+			if err != nil {
+				return nil, err
+			}
+		}
+		klog.Infoln(fmt.Sprintf("package variant %q is doing a pkg update on package revision %q", pv.Name, downstream.Name))
+		downstreams[i], err = r.updateDraft(ctx, downstream, upstream)
 		if err != nil {
 			return nil, err
 		}
 	}
-
-	return r.updateDraft(ctx, downstream, upstream)
+	return downstreams, nil
 }
 
-func (r *PackageVariantReconciler) getDownstreamPR(pv *api.PackageVariant,
-	prList *porchapi.PackageRevisionList) *porchapi.PackageRevision {
+// If there are any drafts that are owned by us and match the target package
+// revision, return them all. If there are no drafts, return the latest published
+// package revision owned by us.
+func (r *PackageVariantReconciler) getDownstreamPRs(ctx context.Context,
+	pv *api.PackageVariant,
+	prList *porchapi.PackageRevisionList) []*porchapi.PackageRevision {
 	downstream := pv.Spec.Downstream
 
 	var latestPublished *porchapi.PackageRevision
+	var drafts []*porchapi.PackageRevision
 	// the first package revision number that porch assigns is "v1",
 	// so use v0 as a placeholder for comparison
 	latestVersion := "v0"
 
 	for _, pr := range prList.Items {
-		// look for the downstream package in the target repo
+		owned := r.hasOurOwnerReference(pv, pr.ObjectMeta.OwnerReferences)
```

Comment:
Yeah as we discussed offline, I agree that we should eventually have a way to find the upstream packagerevision without having to list all the package revisions (imo this could be a blocker for cross-namespace cloning). But since we don't have that yet, this controller needs to list all of them. 

When this controller no longer needs to list all the packagerevisions, we can update this controller to only list certain packagerevisions. I added a TODO for that. 

A caveat here is if the `adoptionPolicy` field is set to `adoptExisting`, we will still need to list all the packagerevisions, so we can figure out which ones this controller needs to adopt. A mechanism to filter out packagerevisions by repo and/or package name would be helpful for that.